src/Hl7.Fhir.Base/FhirPath/Expressions/EchoVisitor.cs

```diff
@@ -0,0 +1,301 @@
+        }
+
+        private void OutputSubToken(SubToken subtoken)
+        {
+            if (subtoken == null) return;
+            OutputPrecedingTokens(subtoken);
+            _result.Append($"{subtoken.Value}");
+            OutputTrailingTokens(subtoken);
+        }
+
+        private void OutputPrecedingTokens(SubToken subtoken)
+        {
+            if (subtoken == null) return;
+            if (subtoken.leadingWhitespace?.Any() == true)
+                _result.Append(System.String.Join("", subtoken.leadingWhitespace.Select(ws => ws.ToString())));
+        }
+        private void OutputTrailingTokens(SubToken subtoken)
+        {
+            if (subtoken == null) return;
+            if (subtoken.trailingWhitespace?.Any() == true)
+                _result.Append(System.String.Join("", subtoken.trailingWhitespace.Select(ws => ws.ToString())));
+        }
+        #endregion
+
+        public override StringBuilder VisitConstant(ConstantExpression expression)
+        {
+            OutputPrecedingTokens(expression);
+            if (expression is IdentifierExpression identifier)
+            {
+                OutputIdentifierName(identifier.Value);
+                OutputTrailingTokens(expression);
+                return _result;
+            }
+            var t = Fhir.Serialization.PrimitiveTypeConverter.ConvertTo<string>(expression.Value);
+
+            switch (expression.ExpressionType.Name)
+            {
+                case "Any":
+                    _result.Append($"{t}");
+                    break;
+                case "Boolean":
+                    _result.Append($"{t}");
+                    break;
+                case "Code":
+                    _result.Append($"{t}");
+                    break;
+                case "Concept":
+                    _result.Append($"{t}");
+                    break;
+                case "Date":
+                    _result.Append($"@{t}");
+                    break;
+                case "DateTime":
+                    if (t.Contains('T'))
+                        _result.Append($"@{t}");
+                    else
+                        _result.Append($"@{t}T");
+                    break;
+                case "Decimal":
+                    _result.Append($"{t}");
+                    break;
+                case "Integer":
+                    _result.Append($"{t}");
+                    break;
+                case "Long":
+                    _result.Append($"{t}");
+                    break;
+                case "Quantity":
+                    if (expression.Value is P.Quantity q)
+                    {
+                        _result.Append($"{q?.Value}");
+                        OutputSubToken(expression.Unit);
+                    }
+                    else
+                    {
+                        _result.Append($"{t}");
+                    }
+                    break;
+                case "String":
+                    _result.Append($"'{t}'");
```

Comment:
This line should really be handling the delimiting of the strings too.
Should the `Functions.StringOperators.EscapeJson` be the appropriate function to use there?
```
_result.Append("'" + Functions.StringOperators.EscapeJson(t) + "'"); 
```

Reply:
Yeah, if the goal of this visitor is to roundtrip to the original fhirpath as much as possible, it should escape stuff. Not easy to really roundtrip (as you cannot distinguish between allowed unicode characters and escaped unicodes anymore), but that would be good enough.  So, it should "undo" the parsing of the string.

If this were C# we could leverage Roslyn:

```
  return SyntaxFactory.LiteralExpression(SyntaxKind.StringLiteralExpression, SyntaxFactory.Literal(input)).ToFullString();
```

But I don't know if that's good enough.